tests/MongoDB.Driver.Core.Tests/Jira/CSharp3173Tests.cs

```diff
@@ -0,0 +1,358 @@
+﻿/* Copyright 2020-present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Linq;
+using System.Net;
+using System.Threading;
+using System.Threading.Tasks;
+using FluentAssertions;
+using MongoDB.Bson;
+using MongoDB.Bson.TestHelpers.XunitExtensions;
+using MongoDB.Driver.Core.Clusters;
+using MongoDB.Driver.Core.Clusters.ServerSelectors;
+using MongoDB.Driver.Core.Configuration;
+using MongoDB.Driver.Core.ConnectionPools;
+using MongoDB.Driver.Core.Connections;
+using MongoDB.Driver.Core.Events;
+using MongoDB.Driver.Core.Helpers;
+using MongoDB.Driver.Core.Servers;
+using MongoDB.Driver.Core.WireProtocol.Messages;
+using MongoDB.Driver.Core.WireProtocol.Messages.Encoders;
+using Moq;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace MongoDB.Driver.Core.Tests.Jira
+{
+    public class CSharp3173Tests
+    {
+        private readonly static ClusterConnectionMode __clusterConnectionMode = ClusterConnectionMode.Sharded;
+        private readonly static ClusterId __clusterId = new ClusterId();
+        private readonly static EndPoint __endPoint1 = new DnsEndPoint("localhost", 27017);
+        private readonly static EndPoint __endPoint2 = new DnsEndPoint("localhost", 27018);
+        private readonly static TimeSpan __heartbeatInterval = TimeSpan.FromMilliseconds(200);
+        private readonly static ServerId __serverId1 = new ServerId(__clusterId, __endPoint1);
+        private readonly static ServerId __serverId2 = new ServerId(__clusterId, __endPoint2);
+
+        [Theory]
+        [ParameterAttributeData]
+        public void Ensure_command_network_error_before_hadnshake_is_correctly_handled([Values(false, true)] bool async, [Values(false, true)] bool streamable)
```

Comment:
Logic:
1. we launch heartbeat
2. we select the server (27017)
3. Then we pause the isMaster response for this server until the GetChannel will trigger network exception
4. Then we cancel the current isMaster paused check. After that we come to this line: https://github.com/DmitryLukyanov/mongo-csharp-driver/blob/6ed472e10c07b86f098691ac7ad80dd7e91296e7/src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs#L331 So, the canceled description is not published

NOTE: This case is not covered by the test since I didn't find the way to mock it, but it's also possible that we canceled the master check after line 331. In this case, we will throw cancellation exception in this line: https://github.com/DmitryLukyanov/mongo-csharp-driver/blob/6ed472e10c07b86f098691ac7ad80dd7e91296e7/src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs#L383